src/main/java/org/apache/sling/jcr/repoinit/impl/NodePropertiesVisitor.java

```diff
@@ -70,29 +76,116 @@ private static boolean needToSetProperty(Node n, PropertyLine line) throws Repos
+            }
+            pRelPath = String.format("%s/%s", subTreePath, name);
+        }
+        return pRelPath;
+    }
+
+    /**
+     * Lookup the authorizables for the given ids
+     * @param session the jcr session
+     * @param ids delimited list of authorizable ids
+     * @return iterator over the found authorizables
+     */
+    @NotNull
+    private static Iterable<Authorizable> getAuthorizables(@NotNull Session session, @NotNull String ids) throws RepositoryException {
+        List<Authorizable> authorizables = new ArrayList<>();
+        for (String id : Text.explode(ids, ID_DELIMINATOR)) {
+            Authorizable a = UserUtil.getAuthorizable(session, id);
+            if (a == null) {
+                throw new PathNotFoundException("Cannot resolve path of authorizable with id '" + id + "'.");
+            }
+            authorizables.add(a);
+        }
+        return authorizables;
+    }
+
     @Override
     public void visitSetProperties(SetProperties sp) {
         for (String nodePath : sp.getPaths()) {
             try {
-                log.info("Setting properties on nodePath '{}'", nodePath);
-                Node n = session.getNode(nodePath);
-                for (PropertyLine pl : sp.getPropertyLines()) {
-                    final String pName = pl.getPropertyName();
-                    final PropertyLine.PropertyType pType = pl.getPropertyType();
-                    final List<Object> values = pl.getPropertyValues();
-                    final int type = PropertyType.valueFromName(pType.name());
-                    if (needToSetProperty(n, pl)) {
-                        if (values.size() > 1) {
-                            Value[] pValues = convertToValues(values);
-                            n.setProperty(pName, pValues, type);
+                if (nodePath.startsWith(PATH_AUTHORIZABLE)) {
+                    // special case for setting properties on authorizable
+                    int lastHashIndex = nodePath.lastIndexOf(SUBTREE_DELIMINATOR);
+                    if (lastHashIndex == -1) {
+                        throw new IllegalStateException("Invalid format of authorizable path: # deliminator expected.");
+                    }
+                    String ids = nodePath.substring(PATH_AUTHORIZABLE.length(), lastHashIndex);
+                    String subTreePath = nodePath.substring(lastHashIndex + 1);
+                    for (Authorizable a : getAuthorizables(session, ids)) {
+                        log.info("Setting properties on authorizable '{}'", a.getID());
+                        for (PropertyLine pl : sp.getPropertyLines()) {
+                            final String pName = pl.getPropertyName();
+                            final String pRelPath = toRelPath(subTreePath, pName);
+                            final List<Object> values = pl.getPropertyValues();
+                            if (needToSetProperty(a, pRelPath, pl.isDefault())) {
```

Comment:
this looks very similar to the code that also uses needToSetProperty a few lines down, and the many nested if/else are a bit hard to read. Maybe something can be factored out for readability?

Reply:
Well, yes, the logic for determining if a property should be set and setting that value is similar for authorizable vs node.  However, the differences appear to be in all the important places.  Maybe we could try to extract some common handling by passing around a few callback functions to handle the those slight differences, but to me that abstraction makes the code less readable, not more readable.

I never found nested loops and conditional statements difficult to read, but I will extract those blocks into methods to address your concerns.  That should also resolve the last "code smell" reported by sonar as well.

It seems that the alternative to nested loops and conditions is a extracting a bunch of small methods that you must jump around the source code to read rather than having all the logic in one place, so I'm not sold that those "Cognitive Complexity" calculations from sonar have much real world value.